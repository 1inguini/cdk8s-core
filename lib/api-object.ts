import { Construct } from '@aws-cdk/core';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';
import { removeEmpty } from './util';

export interface ApiObjectProps {
  /**
   * The API object spec.
   */
  readonly spec?: any;

  /**
   * Data associated with the resource.
   */
  readonly data?: any;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   * Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ObjectmetaMetaV1;
}

export class ApiObject extends Construct {

  /**
   * The app-unique name of the object.
   *
   * The name is allocated based on the path of the object construct within the
   * construct tree.
   */
  public readonly name: string;

  public readonly apiVersion: string;
  public readonly kind: string;

  private readonly metadata: ObjectmetaMetaV1;

  constructor(scope: Construct, id: string, private readonly props: ApiObjectProps) {
    super(scope, id);

    this.kind = props.kind;
    this.apiVersion = props.apiVersion;

    this.metadata = props.metadata || {};
    this.metadata.name = this.metadata.name || this.node.uniqueId.toLowerCase();
    this.name = this.metadata.name;
  }

  /**
   * Renders the object to Kubernetes config.
   */
  public render(): any {
    return removeEmpty({
      apiVersion: this.props.apiVersion,
      kind: this.props.kind,
      metadata: this.metadata,
      data: this.props.data,
      spec: this.props.spec
    });
  }
}

