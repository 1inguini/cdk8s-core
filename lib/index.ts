export * from './service';
export * from './deployment';
export * from './resource';
export * from './stack';
