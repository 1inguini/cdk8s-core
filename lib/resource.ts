import { Construct } from '@aws-cdk/core';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';

export interface KubResourceProps {
  /**
   * The resource API spec.
   */
  readonly spec?: any;

  /**
   * Data associated with the resource.
   */
  readonly data?: any;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   s* Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ObjectmetaMetaV1;
}

export class KubResource extends Construct {

  /**
   * The app-unique name of the resource.
   *
   * The name is allocated based on the path of the resource construct within the
   * construct tree.
   */
  public readonly name: string;

  private readonly metadata: ObjectmetaMetaV1;

  constructor(scope: Construct, id: string, private readonly props: KubResourceProps) {
    super(scope, id);

    this.metadata = props.metadata || {};
    this.metadata.name = this.metadata.name || this.node.uniqueId.toLowerCase();
  }

  /**
   * Renders the resource to Kubernetes config.
   */
  public render(): any {
    return removeEmpty({
      apiVersion: this.props.apiVersion,
      kind: this.props.kind,
      metadata: this.metadata,
      data: this.props.data,
      spec: this.props.spec
    });
  }
}

function removeEmpty(obj: any) {
  Object.entries(obj).forEach(([ key, value ]) => {
    if (!Array.isArray(obj) && typeof value === 'object') {
      removeEmpty(value);
    }
    else if (value == null) {
      delete obj[key];
    }
  });
  return obj;
};