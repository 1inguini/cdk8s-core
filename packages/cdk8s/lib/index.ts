export * from './service';
export * from './deployment';
export * from './api-object';
export * from './chart';
export * from './horizontal-pod-autoscaler';
export * from './ingress';