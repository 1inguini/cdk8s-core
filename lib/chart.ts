import { Construct, ISynthesisSession } from '@aws-cdk/core';
import fs = require('fs');
import path = require('path');
import { ApiObject } from './api-object';
const yaml = require('js-yaml');


export class Chart extends Construct {

  /**
   * The name of the stack's YAML file as emitted into the cloud assembly
   * directory during synthesis.
   */
  public readonly assemblyFileName: string;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.assemblyFileName = `${this.node.uniqueId}.k8s.yaml`;
  }

  protected synthesize(session: ISynthesisSession) {
    const resources = new Array<any>();

    for (const resource of this.node.findAll().filter(x => x instanceof ApiObject)) {
      resources.push((resource as ApiObject).render());
    }

    const doc = resources.map(r => yaml.dump(r, { skipInvalid: true })).join('---\n');
    fs.writeFileSync(path.join(session.assembly.outdir, this.assemblyFileName), doc);
  }
}
