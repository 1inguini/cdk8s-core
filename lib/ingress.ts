import { ApiObject } from './api-object';
import { Construct } from '@aws-cdk/core';
import { IngressspecExtensionsV1Beta1 } from './.generated/ingressspec-extensions-v1beta1';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';

export interface IngressObjectOptions {
  readonly spec: IngressspecExtensionsV1Beta1;
  readonly metadata: ObjectmetaMetaV1;
}

export class IngressObject extends ApiObject {
  constructor(scope: Construct, id: string, options: IngressObjectOptions) {
    super(scope, id, {
      apiVersion: 'extensions/v1beta1',
      kind: 'Ingress',
      spec: options.spec,
      metadata: options.metadata
    });
  }
}