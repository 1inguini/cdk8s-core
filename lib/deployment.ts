import { KubResource } from './resource';
import { Construct } from '@aws-cdk/core';
import { DeploymentspecAppsV1 } from './.generated/deploymentspec-apps-v1';

export class KubDeployment extends KubResource {
  constructor(scope: Construct, ns: string, spec: DeploymentspecAppsV1) {
    super(scope, ns, {
      apiVersion: 'apps/v1',
      kind: 'Deployment',
      spec: spec
    });
  }
}
