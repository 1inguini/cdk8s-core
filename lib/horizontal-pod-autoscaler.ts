import { KubResource } from './resource';
import { Construct } from '@aws-cdk/core';
import { HorizontalpodautoscalerAutoscalingV2Beta1 } from './.generated/horizontalpodautoscaler-autoscaling-v2beta1';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';

export interface KubHorizontalPodAutoscalerOptions {
  readonly spec: HorizontalpodautoscalerAutoscalingV2Beta1;
  readonly metadata?: ObjectmetaMetaV1;
}

export class KubHorizontalPodAutoscaler extends KubResource {
  constructor(scope: Construct, ns: string, options: KubHorizontalPodAutoscalerOptions) {
    super(scope, ns, {
      apiVersion: 'apps/v1',
      kind: 'Deployment',
      spec: options.spec,
      metadata: options.metadata
    });
  }
}
