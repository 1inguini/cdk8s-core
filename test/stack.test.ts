import { Stack, KubResource } from '../lib';
import { App, Construct } from '@aws-cdk/core';
import { synth } from './util';

test('empty stack', () => {
  // GIVEN
  const app = new App();

  // WHEN
  const stack = new Stack(app, 'empty');

  // THEN
  expect(synth(stack)).toMatchSnapshot();
});

test('output includes all synthesized resources', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'test');

  // WHEN
  new KubResource(stack, 'Resource1', { kind: 'Resource1', apiVersion: 'v1' });
  new KubResource(stack, 'Resource2', { kind: 'Resource2', apiVersion: 'v1' });
  new KubResource(stack, 'Resource3', { kind: 'Resource3', apiVersion: 'v1' });

  // also, subscopes
  const scope = new Construct(stack, 'scope');
  new KubResource(scope, 'Resource1', { kind: 'Resource1', apiVersion: 'v1' });
  new KubResource(scope, 'Resource2', { kind: 'Resource2', apiVersion: 'v1' });

  // THEN
  expect(synth(stack)).toMatchSnapshot();
});