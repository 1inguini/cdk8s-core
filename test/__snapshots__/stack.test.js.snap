// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty stack 1`] = `
Array [
  null,
]
`;

exports[`output includes all synthesized resources 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Resource1",
    "metadata": Object {
      "name": "testresource156b797a3",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Resource2",
    "metadata": Object {
      "name": "testresource2fa260d39",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Resource3",
    "metadata": Object {
      "name": "testresource3e1820168",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Resource1",
    "metadata": Object {
      "name": "testscoperesource17339213a",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Resource2",
    "metadata": Object {
      "name": "testscoperesource2ea6a0cff",
    },
  },
]
`;
